Multiple updates to exchange_rate sample dapp for Canister HTTP feature.

qj-yu opened the pull request:
**Overview**

This PR include multiple changes to the `exchange_rate` sample dapp. 

- Updated `http_request` call to be in line with newer `transform` callback function format defined in replica.
- Added motoko version of implementation.
- Added security headers requested by security team.

krpeacock commented on rust/exchange_rate/README.md:
```diff
@@ -57,7 +57,9 @@ triggering remote HTTP calls again.
 
 ## Deploy the canister locally
+- `cargo install cargo-audit` to install vulnerability inspection tool
+- `npm i svelte` to prepare needed UI canister dependencies
```
Generally, installing an individual package won't be necessary in a step like this, since Svelte is included in the `package.json`. If this issue came up during your process, I'd imagine that a simple `npm install` is probably all that was needed

qj-yu replied:
Got it.  I will replace `npm i svelte` to `npm install` to catch all. Thanks!

qj-yu replied:
updated

sesi200 commented on motoko/exchange_rate/canister_ids.json:
```diff
@@ -0,0 +1,11 @@
+{
+  "__Candid_UI": {
+    "ic": "a4gq6-oaaaa-aaaab-qaa4q-cai"
+  },
+  "exchange_rate": {
+    "ic": "za3g3-aiaaa-aaaao-qaaca-cai"
+  },
+  "exchange_rate_assets": {
+    "ic": "zh2ap-nqaaa-aaaao-qaacq-cai"
+  }
+}
```
`dfx deploy --network ic` will try to deploy to these canisters. I think it's best to delete this file

qj-yu replied:
Hi @sesi200, I thought regular developers can't deploy to these canister_ids even if they know the id and have the source code, as they are not canister controllers. 

On the other hand, we are trying to announce the canister_ids in public documentation, so these are not hidden info even if I remove them from here. 

This file is mostly to make future deployments easier. If you strongly prefer to remove them from here, I'd be happy to remove.

sesi200 replied:
Looks like I didn't explain myself well enough, sorry.

> I thought regular developers can't deploy to these canister_ids even if they know the id and have the source code, as they are not canister controllers.

This is correct.

> On the other hand, we are trying to announce the canister_ids in public documentation, so these are not hidden info even if I remove them from here.

IMO (which may be wrong 😄) noone will discover the canister IDs of the example if you only put them into this file. I would actually put the links to the example deployment into the README

> This file is mostly to make future deployments easier.

I totally get that. My main concern is that if someone else (who isn't a controller of these canisters) tries to deploy this to mainnet, then it will _also_ try to deploy to these canister IDs and give `403 permission denied`. And if a beginner runs into this they won't think of checking `canister_ids.json` because they don't know how things work behind the scenes.

I'm ok leaving this in here, but I think it should at least be renamed (e.g. to `canister_ids.json_example_deployment`) so that a simple `dfx deploy --network ic` can work

qj-yu replied:
Thanks for the detailed explanation @sesi200 !  I renamed `canister_ids.json` to `canister_ids.json.example` as a middle ground.

sesi200 commented on motoko/exchange_rate/dfx.json:
```diff
@@ -0,0 +1,29 @@
+      "package": "exchange_rate",
+      "type": "motoko",
+      "main": "src/Main.mo",
+      "wasm": "target/wasm32-unknown-unknown/release/exchange_rate.wasm"
```
This looks like something from a Rust project, not Motoko. Is this right?

qj-yu replied:
Right, I copied the `dfx.json` from rust folder to motoko.  Will remove this line.

qj-yu replied:
This is cleaned up.

sesi200 commented on rust/exchange_rate/README.md:
```diff
@@ -57,7 +57,9 @@ triggering remote HTTP calls again.
+- `npm install` to prepare needed UI canister dependencies
 - `dfx start --enable-canister-http` to start a local IC instance with Canister HTTP feature enabled
-- `dfx deploy --with-cycles 100000000000` to deploy the `exchange_rate` and `exchange_rate_assets`
+- `dfx deploy --with-cycles=40000000000000` to deploy the `exchange_rate` and `exchange_rate_assets`
```
Motoko version currently has no README at all.

qj-yu replied:
Will create one.

qj-yu replied:
Have created `motoko` README and checked in.

sesi200 commented on rust/exchange_rate/canister_ids.json:
```diff
@@ -0,0 +1,11 @@
+{
+  "__Candid_UI": {
+    "ic": "a4gq6-oaaaa-aaaab-qaa4q-cai"
+  },
+  "exchange_rate": {
+    "ic": "zs5rc-myaaa-aaaao-qaaba-cai"
+  },
+  "exchange_rate_assets": {
+    "ic": "zv4xw-baaaa-aaaao-qaabq-cai"
+  }
+}
```
`dfx deploy --network ic` will try to deploy to these canisters. I think it's best to delete this file

qj-yu replied:
Same as above

sesi200 reviewed (changes requested):
What version of dfx are you using to test this? Both `master` and `0.11.2` produce the following error for the Motoko version when running `dfx deploy`:
```
The post-build step failed for canister 'ryjl3-tyaaa-aaaaa-aaaba-cai' (exchange_rate_assets) with an embedded error: Failed to assert source paths.: Unable to determine canonical location of asset source path /Users/ssiff/Desktop/examples/motoko/exchange_rate/src/frontend/public: No such file or directory (os error 2)
```

qj-yu commented:
> What version of dfx are you using to test this? Both `master` and `0.11.2` produce the following error for the Motoko version when running `dfx deploy`:
> 
> ```
> The post-build step failed for canister 'ryjl3-tyaaa-aaaaa-aaaba-cai' (exchange_rate_assets) with an embedded error: Failed to assert source paths.: Unable to determine canonical location of asset source path /Users/ssiff/Desktop/examples/motoko/exchange_rate/src/frontend/public: No such file or directory (os error 2)
> ```

I'll submit a commit with `deploy.sh` script, which copies frontend assets from `rust` folder to `motoko`.   I didn't want to duplicate that portion of code.

qj-yu commented:
Thanks for reviewing @sesi200 !  I'll follow up with fixes to your comments shortly.

qj-yu commented:
Hi @sesi200 , I've updated this PR and address above code review comments.   In summary, the new updates are:

* Added `README.md` for `motoko` dapp
* Added `deploy.sh` scripts for both `rust` and `motoko` dapp, and tested locally that they both worked (with `dfx 0.12.0-beta.3` dependency though)
* Renamed `canister_ids.json` to `canister_ids.json.example` to prevent encountering `403 Permission Denied` error

Both dapps are tested locally with `dfx 0.12.0-beta.3` and running live on mainnet. Could you help me with another look? Thanks!

sesi200 commented on motoko/exchange_rate/cleanup.sh:
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
+
+# Call the script with deploy.sh {testnet} {git_revision}
```
```suggestion
# Call the script with cleanup.sh {network}
```

sesi200 commented on motoko/exchange_rate/deploy.sh:
```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
+
+# Call the script with deploy.sh {testnet} {git_revision}
```
```suggestion
# Call the script with deploy.sh {network}
```

sesi200 commented on rust/exchange_rate/cleanup.sh:
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
+
+# Call the script with deploy.sh {testnet} {git_revision}
```
```suggestion
# Call the script with cleanup.sh {network}
```

sesi200 commented on rust/exchange_rate/deploy.sh:
```diff
@@ -0,0 +1,32 @@
+#!/bin/bash
+
+# Call the script with deploy.sh {testnet} {git_revision}
```
```suggestion
# Call the script with deploy.sh {network}
```

sesi200 commented on rust/exchange_rate/dfx.json:
```diff
@@ -21,9 +21,13 @@
-      "bind": "127.0.0.1:8000",
+      "bind": "127.0.0.1:4943",
       "type": "ephemeral"
+    },
+    "ic": {
+      "bind": "https://ic0.app",
+      "type": "persistent"
```
```suggestion
```
shouldn't be necessary

sesi200 reviewed (changes requested):
Here's a few more minor changes, then we can merge this

qj-yu commented:
Thanks for suggestion @sesi200 ! All valid, and adopted.

sesi200 reviewed (approved):
Thanks a lot, QJ!
